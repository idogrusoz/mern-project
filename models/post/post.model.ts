import { model } from "mongoose";
import PostSchema from "./post.schema";
import { IPostModel, PostDocument } from "./post.types";

export const PostModel: IPostModel = model<PostDocument, IPostModel>("Post", PostSchema);
