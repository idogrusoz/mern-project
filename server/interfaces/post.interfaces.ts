export interface IPostInterface {
    author: PostAuthor;
    textContent: string;
    style: PostStyle;
    likes: string[];
    createdAt?: Date;
    updatedAt?: Date;
}

export interface PostStyle {
    fontFamily: string;
    fontSize: number;
    fontWeight: number;
    backgroundColor: string;
    textAlign: string;
    color: string;
}

export interface PostAuthor {
    user_id: string;
    userName: string;
    image: string;
}

export interface BasePostDocument extends IPostInterface {
    _id: string;
}
