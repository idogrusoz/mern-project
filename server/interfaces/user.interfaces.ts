export interface BaseUser {
    firstName: string;
    lastName: string;
    displayName: string;
    email: string;
    userName: string;
    image?: string;
}

export interface UserInterface extends BaseUser {
    password: string;
}

export interface ResponseUser extends BaseUser {
    _id: string;
}

export interface SignInUser {
    email: string;
    password: string;
}

export interface SearchedUser {
    _id: string;
    userName: string;
    image: string;
    displayName: string;
}
