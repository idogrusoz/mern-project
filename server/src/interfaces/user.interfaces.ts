export interface UserInterface {
    firstName: string;
    lastName: string;
    displayName: string;
    email: string;
    password: string;
}

export interface ServiceResponse<T> {
    statusCode: number;
    error: boolean;
    message?: string;
    data?: T;
}
