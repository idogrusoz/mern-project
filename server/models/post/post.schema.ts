import { PostDocument } from "./post.types";
import { Schema } from "mongoose";

const PostSchema = new Schema({
    author: {
        user_id: {
            type: String,
            indexes: true,
            required: true,
        },
        userName: {
            type: String,
        },
        image: {
            type: String,
        },
    },
    textContent: {
        type: String,
    },
    style: {
        fontFamily: {
            type: String,
        },
        fontSize: {
            type: Number,
        },
        fontWeight: {
            type: Number,
        },
        backgroundColor: {
            type: String,
        },
        textAlign: {
            type: String,
        },
        color: {
            type: String,
        },
    },
    likes: {
        type: "array",
        items: { type: "string", uniqueItems: true },
    },
    createdAt: {
        type: Date,
    },
    updatedAt: {
        type: Date,
    },
});

PostSchema.statics.findByLikes = function (userId: string) {
    var Post = this;
    return Post.find({ likes: userId });
};

PostSchema.pre<PostDocument>("save", function (this: PostDocument, next) {
    var post = this;
    if (post.isModified()) {
        post.updatedAt = new Date();
        return next();
    }
    post.createdAt = new Date();
    return next();
});

export default PostSchema;
