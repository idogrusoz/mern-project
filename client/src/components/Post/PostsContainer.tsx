import { CircularProgress } from "@material-ui/core";
import React, { FunctionComponent, useEffect } from "react";
import { BasePostDocument } from "../../../../interfaces/post.interfaces";
import PostDisplay from "./PostDisplay";
import EmptyProfile from "../Profile/EmptyProfile";

export type PostContainerProps = {
    posts: BasePostDocument[];
    fetchLikedPosts: () => Promise<void>;
    fetchFeed: () => Promise<void>;
    fetchUserPosts: () => Promise<void>;
    activeTab: "feed" | "posts" | "likes";
    isLoading: boolean;
};

const PostsContainer: FunctionComponent<PostContainerProps> = ({
    posts,
    fetchLikedPosts,
    fetchFeed,
    fetchUserPosts,
    activeTab,
    isLoading,
}): JSX.Element => {
    useEffect(() => {
        if (activeTab === "feed") {
            fetchFeed();
        }
        if (activeTab === "posts") {
            fetchUserPosts();
        }
        if (activeTab === "likes") {
            fetchLikedPosts();
        }
    }, [activeTab, fetchFeed, fetchUserPosts, fetchLikedPosts]);
    return isLoading ? (
        <div style={{ display: "flex", justifyContent: "center" }}>
            <CircularProgress />
        </div>
    ) : (
        <>
            {posts.length > 0 ? (
                posts.map((post: BasePostDocument, i: number) => {
                    return <PostDisplay post={post} fetch={fetchLikedPosts} key={i} />;
                })
            ) : (
                <EmptyProfile activeTab={activeTab} />
            )}
        </>
    );
};

export default PostsContainer;
