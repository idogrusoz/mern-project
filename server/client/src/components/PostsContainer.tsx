import { useMediaQuery } from "@material-ui/core";
import React, { FunctionComponent } from "react";
import { IPostInterface } from "../../../interfaces/post.interfaces";
import PostDisplay from "./PostDisplay";

type PostContainerProps = {
    posts: IPostInterface[];
};

const PostsContainer: FunctionComponent<PostContainerProps> = ({ posts }): JSX.Element => {
    const isBigScreen = useMediaQuery("(min-width: 650px)");
    return (
        <div
            style={
                !isBigScreen
                    ? { height: "calc(100vh - 170px)", overflowY: "auto", overflowX: "visible" }
                    : { height: "calc(100vh - 70px)", overflowY: "auto", overflowX: "visible" }
            }
        >
            {posts.map((post: IPostInterface, i: number) => {
                return <PostDisplay post={post} key={i} />;
            })}
        </div>
    );
};

export default PostsContainer;
