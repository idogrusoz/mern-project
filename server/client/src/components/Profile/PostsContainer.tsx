import React, { FunctionComponent } from "react";
import { BasePostDocument } from "../../../../interfaces/post.interfaces";
import PostDisplay from "../PostDisplay";

type PostContainerProps = {
    posts: BasePostDocument[];
    fetch: () => Promise<void>;
};

const PostsContainer: FunctionComponent<PostContainerProps> = ({ posts, fetch }): JSX.Element => {
    return (
        <div>
            {posts.map((post: BasePostDocument, i: number) => {
                return <PostDisplay post={post} fetch={fetch} />;
            })}
        </div>
    );
};

export default PostsContainer;
