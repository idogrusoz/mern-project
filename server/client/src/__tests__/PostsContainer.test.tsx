import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import PostsContainer from "../components/Profile/PostsContainer";
import { IPostInterface } from "../../../interfaces/post.interfaces";

describe("PostsContainer test", () => {
    const posts: IPostInterface[] = [
        {
            likes: [],
            textContent: "",
            author: {
                user_id: "userId",
                image: "",
                userName: "userName",
            },
            style: {
                fontFamily: "Roboto",
                color: "#ffffff",
                backgroundColor: "#ffffff",
                fontWeight: 400,
                fontSize: 40,
                textAlign: "left" as "left" | "right" | "justify" | "center",
            },
        },
    ];
    const wrapper = shallow(<PostsContainer posts={posts} />);
    it("renders without crash", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
