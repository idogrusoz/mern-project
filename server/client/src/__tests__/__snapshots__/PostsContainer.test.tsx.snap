// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsContainer test renders without crash 1`] = `
<div>
  <PostDisplay
    key="0"
    post={
      Object {
        "author": Object {
          "image": "",
          "userName": "userName",
          "user_id": "userId",
        },
        "likes": Array [],
        "style": Object {
          "backgroundColor": "#ffffff",
          "color": "#ffffff",
          "fontFamily": "Roboto",
          "fontSize": 40,
          "fontWeight": 400,
          "textAlign": "left",
        },
        "textContent": "",
      }
    }
  />
</div>
`;
